sklearn/metrics/pairwise.py

```diff
@@ -2178,8 +2221,20 @@ def pairwise_distances(
     elif metric in PAIRWISE_DISTANCE_FUNCTIONS:
         func = PAIRWISE_DISTANCE_FUNCTIONS[metric]
     elif callable(metric):
+        # we got a custom metric and should accept non-float dtypes
+        dtype = X.dtype if hasattr(X, "dtype") else type(X[0])
+        if Y is not None:
+            y_dtype = Y.dtype if hasattr(Y, "dtype") else type(Y[0])
+            if dtype != y_dtype:
+                raise TypeError("X and Y have different dtypes.")
```

Comment:
I agree. I guess it was added to prevent ``check_pairwise_arrays`` from converting to float. But instead I removed this check and made ``check_pairwise_arrays`` preserve the dtype when using a custom metric.